docs/elements.md

```diff
@@ -1697,7 +1703,18 @@ data["params"]["names_from_user"] = [
 <pl-external-grader-variables params-name="names_for_user"></pl-external-grader-variables>
 
 <p>Your code snippet should define the following variables:</p>
-<pl-external-grader-variables params-name="names_from_user"></pl-external-grader-variables>
+<pl-external-grader-variables params-name="names_from_user">
+  <pl-variable name="x" type="numpy array (length $n$)"
+    >Solution to $\mathbf{Ax}=\mathbf{b}$.</pl-variable
+  >
+</pl-external-grader-variables>
+
+<!--
+  The following tag defines an empty list for the given params-name.
+  This is useful for some cases where a parameter must be set to empty to run the external grader.
+  Nothing will be displayed from this tag.
+-->
```

Comment:
It's not clear to me what cases this would be useful in. Do you have an example? It honestly feels super weird to me to have an element that just sets a param to `[]`.

Reply:
This is actually super common, take a look at the Pandas example question that got refactored in this PR. It comes up in cases where you want a student to write a function that the grader code just calls, so there's no need to pass any parameters directly to the student code. I can double check, but I think the external grader requires that be set (even if it's just the empty list).

Just searching our own course, I found 19 examples of setting the empty list in server.py. It would also feel weird not to have this common case covered by HTML definitions, while other cases are.

Reply:
Would it make more sense to update the grader(s) to treat missing or `None` param the same as an empty list?

Reply:
I'm actually not sure. There's some other discussion on this PR about explicitly declaring empty vs silently doing something different with the default. Like if there's a typo in the variable name declaration, the external grader would just pick up None for the variables to be used, which could be frustrating and hard to debug.

I'm not sure who wrote the Python autograder either, but I think @jonatanschroeder wrote some of the other autograders; I'm curious what his thoughts are.

Reply:
I have never worked with Python questions in PL, so I don't have an opinion here. The other autograders I've worked with (including C, Java and some custom graders) don't use global variables for this element to make sense in those contexts.

Reply:
Yeah, the Python autograder is the only one I know of that actually does anything with `names_for_user` or `names_from_user`. I believe @nicknytko himself wrote the Python autograder. In retrospect, I would have been inclined to name this thing `<pl-python-external-grader-variables>` to make it clear that this element *only* does anything useful for the Python autograder. Maybe we can make that clear in the docs?

Reply:
Yeah, I think we can make that a little more clear in the docs. With this change actually, you could use it to display information for other autograders, but they don't use globals. So it's worth being explicit about recommended use cases.

Reply:
I think this change is actually sub-optimal for non-Python graders. If an instructor interprets this element as "oh, this is just a nice way to render a table of variables", they might get bitten by the fact that any updates to the variables in `question.html` won't be updated for students who have already started a question because the data is snapshotted into `params` when the question is generated. This runs counter to how other display elements work. So we might want to call that out in the docs too.

Reply:
I suppose part of my hesitation here stems from the fact that I view elements in `question.html` as being fundamentally related to rendering. Having a specific configuration of an element that doesn't render anything and only adds something to `params` feels very idiosyncratic to me. Frankly, an element that modifies `params` but doesn't also do any submission-handling/validation/grading feels idiosyncratic too, though I acknowledge it's useful for this specific external grader.

Reply:
Yeah, I'm kindof in agreement actually. The way this grader works is a bit cumbersome compared to the others, which is why having this element becomes convenient. If there was some better way to configure the grader outside of setting these lists, I probably wouldn't advocate for this change. But as-written, this change can alleviate a lot of the bloat from having tons of server files that are just defining variables. Our course repo has a _ton_ of these, and being able to remove them will make maintenance significantly easier.

Reply:
Yeah, I gotcha. I acknowledge the convenience, so I won't block this change.

Not to further derail this conversation, but spitballing for future development ideas: if it were possible to populate the initial params statically via a `params` key in `question.json`, or perhaps via YAML frontmatter in a hypothetical `question.md` file, would that be an attractive alternative to you?

Reply:
I think that could be interesting, though the main use case for populating the params dict is to have things that you can use in later templating / question creation, so this edge case benefits heavily from not having to create the extra file (since there's no extra custom logic). It seems like a more limited version of some of the other proposed ways of having a question take parameters from the corresponding assessment.

Maybe not a bad idea but I'm struggling to come up with use cases right now.